src/core/connection.c

```diff
@@ -836,12 +837,40 @@ QuicConnGenerateNewSourceCid(
                 AllocFailure,
                 "Allocation of '%s' failed. (%llu bytes)",
                 "new Src CID",
-                sizeof(QUIC_CID_HASH_ENTRY) + MsQuicLib.CidTotalLength);
+                sizeof(QUIC_CID_SLIST_ENTRY) + MsQuicLib.CidTotalLength);
             QuicConnFatalError(Connection, QUIC_STATUS_INTERNAL_ERROR, NULL);
             return NULL;
         }
-        if (!QuicBindingAddSourceConnectionID(Connection->Paths[0].Binding, SourceCid)) {
-            CXPLAT_FREE(SourceCid, QUIC_POOL_CIDHASH);
+
+        BOOLEAN Collision = FALSE;
+
+        CxPlatDispatchLockAcquire(&MsQuicLib.DatapathLock);
+
+        //
+        // Check whether a new sourceCid collides with any sourceCid which belong to all
+        // the bindings including the ones which this connection doesn't bound.
+        //
+        for (CXPLAT_LIST_ENTRY* Link = MsQuicLib.Bindings.Flink;
+            Link != &MsQuicLib.Bindings;
+            Link = Link->Flink) {
+
+            QUIC_BINDING* Binding =
+                CXPLAT_CONTAINING_RECORD(Link, QUIC_BINDING, Link);
+            QUIC_CONNECTION* Connection1 =
```

Comment:
Can we use a more descriptive name instead of `Connection1`? Like `SearchConnection` or `CollisionConnection`

Reply:
I agree. I will change this.